pkg/actuators/machine/instances.go

```diff
@@ -199,33 +202,52 @@ func getBlockDeviceMappings(blockDeviceMappings []awsproviderv1.BlockDeviceMappi
 		klog.Errorf("No image for given AMI was found")
 		return nil, fmt.Errorf("no image for given AMI not found")
 	}
-	deviceName := describeAMIResult.Images[0].RootDeviceName
-
-	// Only support one blockDeviceMapping
-	volumeSize := blockDeviceMappings[0].EBS.VolumeSize
-	volumeType := blockDeviceMappings[0].EBS.VolumeType
-	blockDeviceMapping := ec2.BlockDeviceMapping{
-		DeviceName: deviceName,
-		Ebs: &ec2.EbsBlockDevice{
-			VolumeSize: volumeSize,
-			VolumeType: volumeType,
-			Encrypted:  blockDeviceMappings[0].EBS.Encrypted,
-		},
-	}
 
-	if aws.StringValue(blockDeviceMappings[0].EBS.KMSKey.ID) != "" {
-		klog.V(3).Infof("Using KMS key ID %q for encrypting EBS volume", *blockDeviceMappings[0].EBS.KMSKey.ID)
-		blockDeviceMapping.Ebs.KmsKeyId = blockDeviceMappings[0].EBS.KMSKey.ID
-	} else if aws.StringValue(blockDeviceMappings[0].EBS.KMSKey.ARN) != "" {
-		klog.V(3).Info("Using KMS key ARN for encrypting EBS volume") // ARN usually have account ids, therefore are sensitive data so shouldn't log the value
-		blockDeviceMapping.Ebs.KmsKeyId = blockDeviceMappings[0].EBS.KMSKey.ARN
-	}
+	rootDeviceFound := false
+	for _, blockDeviceMappingSpec := range blockDeviceMappingSpecs {
+		if blockDeviceMappingSpec.EBS == nil {
+			continue
+		}
+
+		deviceName := blockDeviceMappingSpec.DeviceName
+		volumeSize := blockDeviceMappingSpec.EBS.VolumeSize
+		volumeType := blockDeviceMappingSpec.EBS.VolumeType
+		deleteOnTermination := true
+
+		if blockDeviceMappingSpec.DeviceName == nil {
+			if rootDeviceFound {
+				return nil, errors.New("non root device must have name")
+			}
+			rootDeviceFound = true
+			deviceName = describeAMIResult.Images[0].RootDeviceName
+		}
+
+		blockDeviceMapping := ec2.BlockDeviceMapping{
+			DeviceName: deviceName,
+			Ebs: &ec2.EbsBlockDevice{
+				VolumeSize:          volumeSize,
+				VolumeType:          volumeType,
+				Encrypted:           blockDeviceMappingSpec.EBS.Encrypted,
+				DeleteOnTermination: &deleteOnTermination,
+			},
+		}
+
+		if *volumeType == ec2.VolumeTypeIo1 {
+			blockDeviceMapping.Ebs.Iops = blockDeviceMappingSpec.EBS.Iops
+		}
+
+		if aws.StringValue(blockDeviceMappingSpecs[0].EBS.KMSKey.ID) != "" {
+			klog.V(3).Infof("Using KMS key ID %q for encrypting EBS volume", *blockDeviceMappingSpecs[0].EBS.KMSKey.ID)
```

Comment:
s/ blockDeviceMappingSpecs[0] / blockDeviceMappingSpecs ?

Reply:
thanks, fixed

Reply:
Isn’t this line still using wrong block device https://github.com/openshift/cluster-api-provider-aws/pull/299/files#diff-24a7c87e11cc6413298ddaaa7e7abda0R240?

Reply:
right, fixed